File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2015/Y15D21.kt
package org.gristle.pdxpuzzles.advent.y2015

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.combinatorics.getPairSequence
import org.gristle.pdxpuzzles.utilities.parsing.blankSplit
import org.gristle.pdxpuzzles.utilities.parsing.getIntList
import org.gristle.pdxpuzzles.utilities.parsing.getInts

class Y15D21(input: String) : Day {

    companion object {

        const val ITEM_LIST = """
            Weapons:    Cost  Damage  Armor
            Dagger        8     4       0
            Shortsword   10     5       0
            Warhammer    25     6       0
            Longsword    40     7       0
            Greataxe     74     8       0

            Armor:      Cost  Damage  Armor
            Leather      13     0       1
            Chainmail    31     0       2
            Splintmail   53     0       3
            Bandedmail   75     0       4
            Platemail   102     0       5

            Rings:      Cost  Damage  Armor
            Damage +1    25     1       0
            Damage +2    50     2       0
            Damage +3   100     3       0
            Defense +1   20     0       1
            Defense +2   40     0       2
            Defense +3   80     0       3
        """
    }

    data class Character(val hp: Int, val damage: Int, val armor: Int) {
        fun roundsToKill(other: Character): Int {
            val adjustedDamage = maxOf(1, damage - other.armor)
            return other.hp / adjustedDamage + if (other.hp % adjustedDamage == 0) 0 else 1
        }
    }

    data class Stats(val cost: Int, val damage: Int, val armor: Int) {
        fun toCharacter() = Character(100, damage, armor)
    }

    operator fun Stats.plus(other: Stats): Stats =
        Stats(cost + other.cost, damage + other.damage, armor + other.armor)

    val loadouts: List<Stats> = let {
        val (weaponShop, armorShop, ringShop) = ITEM_LIST.blankSplit()
        val armor = armorShop
            .getInts()
            .chunked(3) { (cost, damage, armor) -> Stats(cost, damage, armor) }
            .plus(Stats(0, 0, 0))
            .toList()
        val rings = ringShop
            .getInts()
            .chunked(4) { (_, cost, damage, armor) -> Stats(cost, damage, armor) }
            .plus(Stats(0, 0, 0))
            .toList()
        weaponShop
            .getInts()
            .chunked(3) { (cost, damage, armor) -> Stats(cost, damage, armor) }
            .flatMap { weaponized -> armor.map { weaponized + it } }
            .flatMap { armored ->
                val doubleRings = rings.dropLast(1)
                    .getPairSequence()
                    .map { (ring, middle) -> ring + middle }
                rings
                    .asSequence()
                    .plus(doubleRings)
                    .map { beRinged -> beRinged + armored }
            }.sortedBy { (cost) -> cost }
            .toList()
    }

    // Generate boss from input stats
    private val boss = input
        .getIntList()
        .let { (hp, damage, armor) -> Character(hp, damage, armor) }

    override fun part1() = loadouts
        .first { stats ->
            val player = stats.toCharacter()
            player.roundsToKill(boss) <= boss.roundsToKill(player)
        }.cost

    override fun part2() = loadouts
        .last { stats ->
            val player = stats.toCharacter()
            player.roundsToKill(boss) > boss.roundsToKill(player)
        }.cost
}

fun main() = Day.runDay(Y15D21::class)

//    Class creation: 7ms
//    Part 1: 91 (1ms)
//    Part 2: 158 (2ms)
//    Total time: 11ms

File: rust/advent/src/bin/2015/y15d21.rs
use std::cmp::max;
use std::ops::Add;
use itertools::Itertools;
use advent::utilities::get_input::get_input;
use utilities::parsing::get_numbers::ContainsNumbers;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

const ITEM_LIST: &str = r"
    Weapons:    Cost  Damage  Armor
    Dagger        8     4       0
    Shortsword   10     5       0
    Warhammer    25     6       0
    Longsword    40     7       0
    Greataxe     74     8       0

    Armor:      Cost  Damage  Armor
    Leather      13     0       1
    Chainmail    31     0       2
    Splintmail   53     0       3
    Bandedmail   75     0       4
    Platemail   102     0       5

    Rings:      Cost  Damage  Armor
    Damage +1    25     1       0
    Damage +2    50     2       0
    Damage +3   100     3       0
    Defense +1   20     0       1
    Defense +2   40     0       2
    Defense +3   80     0       3
";

const HP: u16 = 100;
type Input = (Character, Vec<Stats>);
type Output = u16;

#[derive(Debug, Clone, Copy)]
struct Stats {
    cost: u16,
    damage: u16,
    armor: u16,
}

impl Add for Stats {
    type Output = Self;
    fn add(self, other: Self) -> Self {
        Self {
            cost: self.cost + other.cost,
            damage: self.damage + other.damage,
            armor: self.armor + other.armor,
        }
    }
}

#[derive(Debug)]
struct Character {
    hp: u16,
    damage: u16,
    armor: u16,
}

impl Character {
    fn rounds_to_kill(&self, other: &Self) -> u16 {
        let adjusted_damage =
            max(1, self.damage.checked_sub(other.armor).unwrap_or_default());
        other.hp / adjusted_damage + if other.hp % adjusted_damage == 0 { 0 } else { 1 }
    }
}

impl From<&Stats> for Character {
    fn from(stats: &Stats) -> Self {
        Self {
            hp: HP,
            damage: stats.damage,
            armor: stats.armor,
        }
    }
}
fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(15, 21).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    let (hp, damage, armor) = input.get_numbers().collect_tuple().unwrap();
    let boss = Character { hp, damage, armor };

    let (weapons, armor, rings) = ITEM_LIST.split("\n\n").collect_tuple().unwrap();
    let armor: Vec<Stats> = armor.get_numbers().tuples()
        .map(|(cost, damage, armor)| Stats { cost, damage, armor })
        .chain(std::iter::once(Stats { cost: 0, damage: 0, armor: 0 }))
        .collect();
    let rings: Vec<Stats> = rings.get_numbers().tuples()
        .map(|(_, cost, damage, armor)| Stats { cost, damage, armor })
        .chain(std::iter::once(Stats { cost: 0, damage: 0, armor: 0 }))
        .collect();

    let loadouts = weapons.get_numbers().tuples()
        .map(|(cost, damage, armor)| Stats { cost, damage, armor })
        .flat_map(|weaponized| armor.iter().map(move |&armor| weaponized + armor))
        .flat_map(|armored| {
            let double_rings = rings[..rings.len() - 1].iter()
                .combinations(2)
                .map(|rings| *rings[0] + *rings[1]);
            rings.iter().copied()
                .chain(double_rings)
                .map(move |be_ringed| be_ringed + armored)
        })
        .sorted_unstable_by_key(|stats| stats.cost)
        .collect();
    (boss, loadouts)
}

fn part1((boss, loadouts): &Input) -> Output {
    loadouts.iter()
        .find(|&stats| {
            let player: Character = stats.into();
            player.rounds_to_kill(&boss) <= boss.rounds_to_kill(&player)
        })
        .unwrap()
        .cost
}

fn part2((boss, loadouts): &Input) -> Output {
    loadouts.iter()
        .rev()
        .find(|&stats| {
            let player: Character = stats.into();
            player.rounds_to_kill(&boss) > boss.rounds_to_kill(&player)
        })
        .unwrap()
        .cost
}

#[test]
fn default() {
    let input = get_input(15, 21).unwrap();
    let input = parse_input(&input);
    assert_eq!(91, part1(&input));
    assert_eq!(158, part2(&input));
}

// Input parsed (44μs)
// 1. 91 (7μs)
// 2. 158 (2μs)
// Total: 55μs